### PR TITLE
Feature/osrm table connector

### PR DESCRIPTION
**Pull request by CodeBear801**
# Issue

Implement OSRM table request and response in Golang.  More information could go to: #137 

## Tasklist

 - [x] Implement table request and response
 - [x] add tests 
 - [ ] review

### CODE REVIEW

File: integration/pkg/api/osrm/table/request.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	if v := values.Get(options.KeyDestinations); len(v) > 0 {
+		if destinations, err := genericoptions.ParseElemenets(v); err == nil {
+			r.Destinations = destinations
+		}
+	}
+
+	if v := values.Get(options.KeyAnnotations); len(v) > 0 {
+		if annotations, err := routeoption.ParseAnnotations(v); err == nil {
```
Comment (wangyoucao577) [inline integration/pkg/api/osrm/table/request.go]: The `annotations` of `route` service are different from that of `table` service. `table` service should have its own parser.

File: integration/pkg/api/osrm/table/request.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/Telenav/osrm-backend/integration/pkg/api"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/coordinate"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/genericoptions"
+	routeoption "github.com/Telenav/osrm-backend/integration/pkg/api/osrm/route/options"
```
Comment (wangyoucao577) [inline integration/pkg/api/osrm/table/request.go]: It's strange that import `route/option` here.

File: integration/pkg/api/osrm/table/request.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		// Path
+		Service:     "table",
+		Version:     "v1",
+		Profile:     "driving",
+		Coordinates: coordinate.Coordinates{},
+
+		// Options
+		Sources:      genericoptions.Elements{},
+		Destinations: genericoptions.Elements{},
+		Annotations:  "",
```
Comment (wangyoucao577) [inline integration/pkg/api/osrm/table/request.go]: The default value is `duration`.

File: integration/pkg/api/osrm/table/response.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	Message      string         `json:"message,omitempty"`
+	Sources      []*Source      `json:"sources"`
+	Destinations []*Destination `json:"destinations"`
+	Durations    [][]*float64   `json:"durations"`
+	Distances    [][]*float64   `json:"distances"`
+}
+
+// Source represents as way point object.  All sources will be listed in order.
+type Source struct {
+	route.Waypoint
```
Comment (wangyoucao577) [inline integration/pkg/api/osrm/table/response.go]: If the `Waypoint` shares between `table` and `route` service, let's put it outside `route`. I think just put the definition to the `osrm/` will be fine.

Comment (wangyoucao577) [reply integration/pkg/api/osrm/table/response.go]: Also, the extra `type` is not necessary. You can define it by `Sources []*Waypoint` directly.

**Comment by wangyoucao577** [comment]
I feel our `api` package design still not good enough. E.g. below approach for caller will be better: 
- use `osrm.Coordinate` instead of `coordinate.Coordinate`; 
- use `osrm.RouteRequest` instead of `route.Request`; 

Maybe we can use `internal` package to hide some implementation details, then put all interfaces at the root of `api/osrm` package.

File: integration/pkg/api/osrm/table/response.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package table
+
+import "github.com/Telenav/osrm-backend/integration/pkg/api/osrm/route"
+
+// Response represents OSRM api v1 table response.
+type Response struct {
+	Code         string            `json:"code"`
+	Message      string            `json:"message,omitempty"`
+	Sources      []*route.Waypoint `json:"sources"`
```
Comment (wangyoucao577) [inline integration/pkg/api/osrm/table/response.go]: `table` API uses `route.Waypoint` still looks strange.

**Comment by wangyoucao577** [comment]
This PR shows a lot of un-related changes currently. It looks like you made something wrong on the commit 05dad50. You may need to discard that commit and then make your expected changes again.

File: integration/pkg/api/osrm/table/response.go (hunk 5 above)
Comment (CodeBear801) [reply integration/pkg/api/osrm/table/response.go]: Move Waypoint to package `osrmtype`.  I think common definitions like `coordinate`, `segment` could be moved here.

**Pull request merged**